### PR TITLE
Add metrics for part number in MergeTree in ClickHouse

### PR DESCRIPTION
**Pull request by weeds085490**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add metrics(Parts, PartsActive, PartsInactive) for part number in MergeTree in clickhouse

Detailed description / Documentation draft:

`PartsActive` corresponds to the commited parts.
`PartsInactive` corresponds to the parts which are not in commited status.
`Parts` corresponds to the total number of parts in clickhouse.

We need to deal with three scenarios  to calculate `Parts`、`PartsInactive`、`PartsInactive`.

- [ ] When the table is loaded, recount according to the parts status.
- [ ] When the table is dropped, recount according to the parts status.
- [ ] Handle the transfer of parts state.

> details:
image

### CODE REVIEW

**Comment by sundy-li** [comment]
BTW, it's a summary metric, why not use `system.parts`, it has more detail `metrics` (per database, table ...). 
And it's not a large table，clickhouse exporters (or custom HTTP metrics handler) work good with it.

**Comment by weeds085490** [comment]
> BTW, it's a summary metric, why not use `system.parts`, it has more detail `metrics` (per database, table ...).
> And it's also a table in memory for statistics, clickhouse exporters (or custom HTTP metrics handler) work good with it.

When we are building the monitoring system of clickhouse, we find the number of parts is an important indicator of the overall situation of a cluster. We recommend adding this information to the default metrics just like `PartMutation`、`ReplicatedFetch`. 

It is worth mentioning that it is indeed possible to collect monitoring data from `system.parts` through `TCP` or `HTTP` interface. We have two main considerations:
-  Every time you pull, the number of parts must be recalculated.
-  Since the growth state of parts is an important indicator of whether the cluster is normal, it should be added to the default metrics.

**Comment by weeds085490** [comment]
> BTW, it's a summary metric, why not use `system.parts`, it has more detail `metrics` (per database, table ...).
> And it's also a table in memory for statistics, clickhouse exporters (or custom HTTP metrics handler) work good with it.

We can also get information about `merge` from `system.merges`, but we also need to add `Merge` to the default metrics

**Comment by weeds085490** [comment]
@akuzm  I have added functional tests for the code. Under normal circumstances, the number of Parts is the same as the data aggregated by `system.parts`.  If we delete the table in the atomic database, the deletion of the table is lazy. At this time, the data may be temporarily inconsistent. PTAL.

**Comment by alexey-milovidov** [comment]
@weeds085490 The code looks Ok but the tests deadlocked. I don't see the reason.
Could you please merge/rebase with master?

**Comment by weeds085490** [comment]
> @weeds085490 The code looks Ok but the tests deadlocked. I don't see the reason.
> Could you please merge/rebase with master?


Thanks for reviewing. I have merged with master. PTAL

**Comment by weeds085490** [comment]
> > @weeds085490 The code looks Ok but the tests deadlocked. I don't see the reason.
> > Could you please merge/rebase with master?
> 
> Thanks for reviewing. I have merged with master. PTAL

@alexey-milovidov `DROP TABLE test_table` in my function test  still hung.  It passed when test it locally.

**Comment by weeds085490** [comment]
> > > @weeds085490 The code looks Ok but the tests deadlocked. I don't see the reason.
> > > Could you please merge/rebase with master?
> > 
> > 
> > Thanks for reviewing. I have merged with master. PTAL
> 
> @alexey-milovidov `DROP TABLE test_table` in my function test still hung. It passed when test it locally.

image

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -1265,6 +1280,11 @@ void MergeTreeData::dropAllData()
     LOG_TRACE(log, "dropAllData: removing data from memory.");
 
     DataPartsVector all_parts(data_parts_by_info.begin(), data_parts_by_info.end());
+    DataPartsVector committed_parts = getDataPartsVector({DataPartState::Committed});
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MergeTreeData.cpp:1283]: Doesn't it involve recurdive mutex locking (see a few lines above)?

Comment (weeds085490) [reply src/Storages/MergeTree/MergeTreeData.cpp]: > Doesn't it involve recursive mutex locking (see a few lines above)?

That did cause a deadlock.  If I use an Ordinary database, the deadlock will definitely occur when dropping table. 

@alexey-milovidov  I have fixed it. PTAL.

**Comment by weeds085490** [comment]
@alexey-milovidov  I have modified the code to make the monitored data consistent with the data obtained by the `system.parts` query. 
* The failed test: `Functional stateless tests (release, wide parts enabled)` may not related to my code.
* The failed test: `Functional stateless tests flaky check (address) ` may be caused by tempory temporary inconsistency. 

image

It becomes consistent after the first query possibly caused by `query_log`.

**Comment by alexey-milovidov** [comment]
@weeds085490 
How this temporary inconsistency possible and how it is related to the query_log?

**Comment by weeds085490** [comment]
> @weeds085490
> How this temporary inconsistency possible and how it is related to the query_log?

During the SQL execution of clickhouse, the input data is not based on the same snapshot. For the same filter, we may get two different data in the same SQL when other threads are inserting new data.  Let‘s look at execution log of the failed test: 
`Functional stateless tests flaky check (address)`

image

when executing the SQL:
```sql
SELECT COUNT(1) FROM (SELECT SUM(IF(metric = 'Parts', value, 0)) AS Parts, SUM(IF(metric = 'PartsActive', value, 0)) AS PartsActive, SUM(IF(metric = 'PartsInactive', value, 0)) AS PartsInactive FROM system.metrics) as a INNER JOIN (SELECT toInt64(SUM(1)) AS Parts, toInt64(SUM(IF(active = 1, 1, 0))) AS PartsActive, toInt64(SUM(IF(active = 0, 1, 0))) AS PartsInactive FROM system.parts ) as b USING (Parts,PartsActive,PartsInactive);
```

A new part: `202012_23_23_0 ` is inserted. So the different input stream may get different parts info.

Then I do another  test：

Reducing the `flush_interval_milliseconds` and `collect_interval_milliseconds` of metric_log.  
I can reproduce this short-term inconsistency stably.

image

If I close metric_log, then this short-term inconsistency will never reproduce.


@alexey-milovidov  PTAL.

**Comment by weeds085490** [comment]
> > @weeds085490
> > How this temporary inconsistency possible and how it is related to the query_log?
> 
> During the SQL execution of clickhouse, the input data is not based on the same snapshot. For the same filter, we may get two different data in the same SQL when other threads are inserting new data. Let‘s look at execution log of the failed test:
> `Functional stateless tests flaky check (address)`
> 
> image
> 
> when executing the SQL:
> 
> ```sql
> SELECT COUNT(1) FROM (SELECT SUM(IF(metric = 'Parts', value, 0)) AS Parts, SUM(IF(metric = 'PartsActive', value, 0)) AS PartsActive, SUM(IF(metric = 'PartsInactive', value, 0)) AS PartsInactive FROM system.metrics) as a INNER JOIN (SELECT toInt64(SUM(1)) AS Parts, toInt64(SUM(IF(active = 1, 1, 0))) AS PartsActive, toInt64(SUM(IF(active = 0, 1, 0))) AS PartsInactive FROM system.parts ) as b USING (Parts,PartsActive,PartsInactive);
> ```
> 
> A new part: `202012_23_23_0 ` is inserted. So the different input stream may get different parts info.
> 
> Then I do another test：
> 
> Reducing the `flush_interval_milliseconds` and `collect_interval_milliseconds` of metric_log.
> I can reproduce this short-term inconsistency stably.
> 
> image
> 
> If I close metric_log, then this short-term inconsistency will never reproduce.
> 
> @alexey-milovidov PTAL.

@alexey-milovidov  Hi, PTAL

**Comment by alexey-milovidov** [comment]
Ok but it means we need to rewrite the test in another way...

**Comment by weeds085490** [comment]
@alexey-milovidov  I have changed the test method. All functional tests have passed. PTAL

**Review by alexey-milovidov — APPROVED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
Sorry, the test is still flaky, I will revert this PR and we need to resubmit it.

File: tests/queries/0_stateless/01600_count_of_parts_metrics.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+FROM system.metrics) as a INNER JOIN
+(SELECT
+    toInt64(SUM(1)) AS Parts,
+    toInt64(SUM(IF(active = 1, 1, 0))) AS PartsActive,
+    toInt64(SUM(IF(active = 0, 1, 0))) AS PartsInactive
+FROM system.parts
+) as b USING (Parts,PartsActive,PartsInactive)"
+
+verify(){
+for _ in $(seq 1 10)
```
Comment (azat) [inline tests/queries/0_stateless/01600_count_of_parts_metrics.sh:22]: The test does not passed after merge - https://clickhouse-test-reports.s3.yandex.net/0/f91626e7ff352d87283e7b8dafed94fd6ef38c8d/fast_test.html#fail1

@weeds085490 can you please take a look? Maybe 10 retries is not enough?

**Comment by alexey-milovidov** [comment]
It failed 4 out of 16 times.

**Comment by weeds085490** [comment]
> It failed 4 out of 16 times.

Ok. Let me take a look

**Comment by alexey-milovidov** [comment]
I will try to revive changes in https://github.com/ClickHouse/ClickHouse/pull/18955

**Comment by alexey-milovidov** [comment]
@weeds085490 See https://github.com/ClickHouse/ClickHouse/pull/19122.

**Pull request merged**